File: neet_code_all_cpp/math_and_geometry/spiral_matrix_2.cpp
#include <cwchar>
#include <iostream>
#include <vector>

class Solution {
    public:
        // AC
        std::vector<std::vector<int>> generateMatrix(int n) {
            std::vector<std::vector<int>> result(n, std::vector<int>(n));
            int left = 0, right = n - 1;
            int top = 0, bottom = n - 1;
            int idx = 1;

            while(top <= bottom && left <= right) {
                for(int i = left; i <= right; ++i) {
                    result[top][i] = idx;
                    ++idx;
                }
                ++top;

                for(int i = top; i <= bottom; ++i) {
                    result[i][right] = idx;
                    ++idx;
                }
                --right;

                if(top <= bottom) {
                    for(int i = right; i >= left; --i) {
                        result[bottom][i] = idx;
                        ++idx;
                    }
                }
                --bottom;

                if(left <= right) {
                    for(int i = bottom; i >= top; --i) {
                        result[i][left] = idx;
                        ++idx;
                    }
                }
                ++left;
            }

            return result;
        }
};

class SolutionAnsKotlin {
    public:
        // AC
        std::vector<std::vector<int>> generateMatrix(int n) {
            std::vector<std::vector<int>> result(n, std::vector<int>(n));
            int left = 0, right = n - 1;
            int top = 0, bottom = n - 1;
            int count = 1;

            while(left <= right && top <= bottom) {
                for(int i = left; i <= right; ++i) {
                    result[top][i] = count;
                    ++count;
                }
                ++top;

                for(int i = top; i <= bottom; ++i) {
                    result[i][right] = count;
                    ++count;
                }
                --right;

                if(left > right || top > bottom) {
                    break;
                }

                for(int i = right; i >= left; --i) {
                    result[bottom][i] = count;
                    ++count;
                }
                --bottom;

                for(int i = bottom; i >= top; --i) {
                    result[i][left] = count;
                    ++count;
                }
                ++left;
            }

            return result;
        }
};

void show_result(std::vector<std::vector<int>>& result) {
    for(auto& v: result) {
        for(auto& w: v) {
            std::cout << w << " ";
        }
        std::cout << '\n';
    }
}

int main(void) {
    Solution s_1;

    int case_1 = 3;
    int case_2 = 1;

    auto res_1 = s_1.generateMatrix(case_1);
    auto res_2 = s_1.generateMatrix(case_2);

    show_result(res_1);
    show_result(res_2);

    SolutionAnsKotlin s_ans_kt;

    auto res_1_kt = s_ans_kt.generateMatrix(case_1);
    auto res_2_kt = s_ans_kt.generateMatrix(case_2);

    show_result(res_1_kt);
    show_result(res_2_kt);
}

File: neet_code_all/src/math_and_geometry/spiral_matrix_2.rs
struct Solution;
impl Solution {
    // AC
    pub fn generate_matrix(n: i32) -> Vec<Vec<i32>> {
        let mut result = vec![vec![0; n as usize]; n as usize];
        let (mut left, mut right) = (0, n - 1);
        let (mut top, mut bottom) = (0, n - 1);

        let mut idx = 1;
        while top <= bottom && left <= right {
            for i in left..=right {
                result[top as usize][i as usize] = idx;
                idx += 1;
            }
            top += 1;

            for i in top..=bottom {
                result[i as usize][right as usize] = idx;
                idx += 1;
            }
            right -= 1;

            if top <= bottom {
                for i in (left..=right).rev() {
                    result[bottom as usize][i as usize] = idx;
                    idx += 1;
                }
            }
            bottom -= 1;

            if left <= right {
                for i in (top..=bottom).rev() {
                    result[i as usize][left as usize] = idx;
                    idx += 1;
                }
            }
            left += 1;
        }

        result
    }
}

// Kotlinの模範解答より
// 自分の回答とほぼ同じ
struct SolutionAnsKotlin;
impl SolutionAnsKotlin {
    // AC
    pub fn generate_matrix(n: i32) -> Vec<Vec<i32>> {
        let mut result = vec![vec![0; n as usize]; n as usize];
        let (mut left, mut right) = (0, n - 1);
        let (mut top, mut bottom) = (0, n - 1);
        let mut count = 1;

        while left <= right && top <= bottom {
            for i in left..=right {
                result[top as usize][i as usize] = count;
                count += 1;
            }
            top += 1;

            for i in top..=bottom {
                result[i as usize][right as usize] = count;
                count += 1;
            }
            right -= 1;

            if left > right || top > bottom {
                break;
            }

            for i in (left..=right).rev() {
                result[bottom as usize][i as usize] = count;
                count += 1;
            }
            bottom -= 1;

            for i in (top..=bottom).rev() {
                result[i as usize][left as usize] = count;
                count += 1;
            }
            left += 1;
        }

        result
    }
}

fn main() {
    let case_1 = 3;
    // [[1,2,3],[8,9,4],[7,6,5]]
    let case_2 = 1;
    // [[1]]

    println!("case_1: {:?}", Solution::generate_matrix(case_1));
    println!("case_2: {:?}", Solution::generate_matrix(case_2));

    println!("case_1: {:?}", SolutionAnsKotlin::generate_matrix(case_1));
    println!("case_2: {:?}", SolutionAnsKotlin::generate_matrix(case_2));
}
